### PR TITLE
HPCC-15582 Revert HPCC-15489 and give slave managers own cache

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: common/thorhelper/roxierow.cpp [hunk 1]
```diff
@@ -386,9 +386,9 @@ IEngineRowAllocator * createCrcRoxieRowAllocator(IRowAllocatorMetaActIdCache * c
 struct AllocatorKey
 {
     IOutputMetaData *meta;
-    unsigned __int64 activityId;
+    unsigned activityId;
     roxiemem::RoxieHeapFlags flags;
-    AllocatorKey(IOutputMetaData *_meta, unsigned __int64 &_activityId, roxiemem::RoxieHeapFlags _flags)
+    AllocatorKey(IOutputMetaData *_meta, unsigned &_activityId, roxiemem::RoxieHeapFlags _flags)
```
Comment (ghalliday) [inline common/thorhelper/roxierow.cpp]: minor: normally activityId wouldn't be a reference parameter

File: roxie/roxiemem/roxiemem.hpp [hunk 2]
```diff
@@ -511,7 +511,7 @@ interface IActivityMemoryUsageMap : public IInterface
 };
 
 extern roxiemem_decl IRowManager *createRowManager(memsize_t memLimit, ITimeLimiter *tl, const IContextLogger &logctx, const IRowAllocatorCache *allocatorCache, bool ignoreLeaks = false, bool outputOOMReports = false);
-extern roxiemem_decl IRowManager *createGlobalRowManager(memsize_t memLimit, memsize_t globalLimit, unsigned numSlaves, ITimeLimiter *tl, const IContextLogger &logctx, const IRowAllocatorCache *allocatorCache, bool ignoreLeaks, bool outputOOMReports);
+extern roxiemem_decl IRowManager *createGlobalRowManager(memsize_t memLimit, memsize_t globalLimit, unsigned numSlaves, ITimeLimiter *tl, const IContextLogger &logctx, const IRowAllocatorCache *allocatorCaches, const IRowAllocatorCache **slaveAllocatorCaches, bool ignoreLeaks, bool outputOOMReports);
```
Comment (ghalliday) [inline roxie/roxiemem/roxiemem.hpp]: trivial: allocatorCaches should lose the trailing s

File: roxie/roxiemem/roxiemem.cpp [hunk 3]
```diff
@@ -4445,15 +4445,16 @@ class CCallbackRowManager : public CChunkingRowManager
 
 class CGlobalRowManager : public CCallbackRowManager
 {
+    const IRowAllocatorCache **slaveAllocatorCaches;
```
Comment (ghalliday) [inline roxie/roxiemem/roxiemem.cpp]: can be deleted.

File: roxie/roxiemem/roxiemem.cpp [hunk 4]
```diff
@@ -4445,15 +4445,16 @@ class CCallbackRowManager : public CChunkingRowManager
...
-        : CCallbackRowManager(_memLimit, _tl, _logctx, _allocatorCache, _ignoreLeaks, _outputOOMReports), numSlaves(_numSlaves)
+    CGlobalRowManager(memsize_t _memLimit, memsize_t _globalLimit, unsigned _numSlaves, ITimeLimiter *_tl, const IContextLogger &_logctx, const IRowAllocatorCache *_allocatorCache, const IRowAllocatorCache **_slaveAllocatorCaches, bool _ignoreLeaks, bool _outputOOMReports)
+        : CCallbackRowManager(_memLimit, _tl, _logctx, _allocatorCache, _ignoreLeaks, _outputOOMReports), numSlaves(_numSlaves), slaveAllocatorCaches(_slaveAllocatorCaches)
     {
         assertex(_globalLimit <= _memLimit);
         globalPageLimit = (unsigned) PAGES(_globalLimit, HEAP_ALIGNMENT_SIZE);
         slaveRowManagers = new CChunkingRowManager * [numSlaves];
         for (unsigned i=0; i < numSlaves; i++)
-            slaveRowManagers[i] = new CSlaveRowManager(i+1, this, _memLimit, _tl, _logctx, _allocatorCache, _ignoreLeaks, _outputOOMReports);
+            slaveRowManagers[i] = new CSlaveRowManager(i+1, this, _memLimit, _tl, _logctx, slaveAllocatorCaches ? slaveAllocatorCaches[i] : nullptr, _ignoreLeaks, _outputOOMReports);
```
Comment (ghalliday) [inline roxie/roxiemem/roxiemem.cpp]: possibly better to say

slaveAllocatorCaches ? slaveAllocatorCaches[i] : allocatorCache

**Comment by ghalliday** [comment]
@jakesmith a few minor comments.

I think the change will work.  I'm not a great fan of passing in the array of slave allocators, but it is relatively clean, so no big objections.

**Comment by jakesmith** [comment]
> I'm not a great fan of passing in the array of slave allocators,

An alternative I supposed would be to either have a :
```
globalRowManger->addSlaveRowManager(<cache>)
```
which create the slaveManager with provided cache at the time.

Or a : 
```
globalRowManager->setSlaveRowManagerCache(unsigned slaveId, <cache>);
```

.. i.e. used default globalRowManager cache, unless this was called to override.

@ghalliday

**Comment by ghalliday** [comment]
I'll merge as-is.  We can possibly think about restructuring the parameters at a later date.

**Pull request merged**